fud2/fud-core/src/exec/plan/egg_planner.rs

```diff
@@ -0,0 +1,260 @@
+use crate::exec::State;
+
+use super::{
+    super::{OpRef, Operation, StateRef},
+    FindPlan, Step,
+};
+use cranelift_entity::PrimaryMap;
+use egg::{
+    define_language, rewrite, Id, Pattern, PatternAst, RecExpr, Rewrite, Runner,
+};
+
+#[derive(Debug, Default)]
+pub struct EggPlanner;
+
+define_language! {
+    /// A language to represent a collection of states.
+    /// For example, if there are 3 states and 2 ops, a term `(root (states x x x) (ops x x))`
+    /// represents the absence of all states and all ops. A term `(root (states 3 x x) (ops x 2))`
```

Comment:
Based on the comments below, I'm confused by the term `(states 3 x x)`. I was expecting the term `(states x x 3)`. Is the latter correct? Or am I missing something.

Reply:
Ah, yes. That comment is now outdated. Thanks. The idea there was to show refs do not have to correspond to indexes in the language. That is relevant now.